dev/NavigationView/NavigationView.cpp

```diff
@@ -3086,17 +3086,78 @@ void NavigationView::UpdateTopNavListViewItemSource(const winrt::IInspectable& i
 
 void NavigationView::UpdateListViewItemSource()
 {
-    if (!m_appliedTemplate)
-    {
-        return;
-    }
-
     auto dataSource = MenuItemsSource();
     if (!dataSource)
     {
         dataSource = MenuItems();
     }
 
+    auto iterable = dataSource.try_as<winrt::IIterable<winrt::IInspectable>>();
+    if (iterable)
+    {
+        // Certain items are disallowed in a NavigationView's items list. Check for them.
+        auto checkItemIsValid = [](const winrt::IInspectable& item)
+        {
+            auto wuxItem = item.try_as<winrt::Windows::UI::Xaml::Controls::NavigationViewItem>();
+            if (wuxItem)
+            {
+                throw winrt::hresult_invalid_argument(L"MenuItems contains a Windows.UI.Xaml.Controls.NavigationViewItem. This control requires that the NavigationViewItems be of type Microsoft.UI.Xaml.Controls.NavigationViewItem.");
+            }
+        };
+
+        // Walk over the entire contents of the items list and validate each item.
+        auto checkListIsValid = [checkItemIsValid](winrt::IIterable<winrt::IInspectable> list)
```

Comment:
I wonder if we could do this in the NavigationViewList's "IsItemItsOwnContainerOverride"? The downside is the call stack of the crash would be pretty deep, but the upside is that we wouldn't have to have a bunch of vectorchanged code just to handle errors.

Reply:
I asked @licanhua if there was a good place to put this kind of per-item validation and didn't think there was any. Looking at IsItemItsOwnContainer that looks like it might be an option. It's a bit of an indirect place to stuff validation. Does it only get triggered during layout? I'd rather have this fail as up front as possible. I totally agree that this is a fair amount of code for what is just a check for a programming error.

Reply:
IsItemItsOwnContainerOverride or PrepareContainerForItemOverride is a good place to go. I didn't remember it before.

---
In reply to: [241144092](https://github.com/Microsoft/microsoft-ui-xaml/pull/69#discussion_r241144092) [](ancestors = 241144092)